README.md
# esp32-cam-websrv

Video streaming webserver for AI Thinker ESP32-CAM boards.

Derived from Espressif's own CameraWebServer example sketch:

https://github.com/espressif/arduino-esp32/tree/master/libraries/ESP32/examples/Camera/CameraWebServer

## Differences from the official CameraWebServer sketch:

* Removed dependence on [arduino-esp32](https://github.com/espressif/arduino-esp32). This application can be built using only the [esp-idf](https://github.com/espressif/arduino-esp32) development framework.
* Removed all face recognition features.
* Added control for the ESP32-CAM's on-board _flash_ LED.
* Added read-only FAT12 partition.
* WIFI SSID and password are now configuration parameters that are read from a configuration file in the FAT12 partition.
* The static webpage data (HTML, CSS and Javascript) are now stored as separate files in the FAT12 parition instead of being hard-coded byte array in a header file.
* A single HTTPD instance (on port 80) serves the static pages, control API, still image and MJPEG stream.
* Multiple clients can view the MJPEG stream simultaneously.
* Added stream framerate control (1 FPS min, 8 FPS max, 4 FPS default).
* Added camera reset button.
* Added custom lightweight ping module to check network connectivity, without the overheads of creating a new session task when using ``esp_ping_*()`` from the ICMP Echo API.

## Build dependency components

* [esp32-camera](https://github.com/espressif/esp32-camera)
* Standard esp-idf components:
    * esp\_event
    * esp\_http\_server
    * esp\_timer
    * esp\_wifi
    * fatfs
    * nvs\_flash
    * vfs

## Setup

Configure, build and flash using the _usual_ IDF tools:

1. Tune configuration

   1. Optionally tweak hard-coded configuration paramerers in ``main/config.h``.

   2. Set WIFI and networking configuration parameters in ``storage/config.cfg``.

      * ``wifi_ssid``: Set to the AP SSID to connect to.
      * ``wifi_pass``: Set to the WPA/2 PSK passphrase.
      * ``ping_host``: Set to IP of host to send ping probes to, or leave blank to disable ping probes.

2. Clean

    ```
    $ idf.py fullclean
    Executing action: fullclean
    Executing action: remove_managed_components
    Done
    ```

3. Set target

    ```
    $ idf.py set-target esp32
    Executing action: set-target
    ...
    -- Configuring done
    -- Generating done
    -- Build files have been written to: /tmp/esp32-cam-websrv/build
    ```

4. Build

    ```
    $ idf.py build
    Executing action: all (aliases: build)
    ...
    Project build complete. To flash, run this command:
    python ../../../../../github/espressif/esp-idf/components/esptool_py/esptool/esptool.py -p (PORT) -b 460800 --before default_reset --after hard_reset --chip esp32  write_flash --flash_mode dio --flash_size detect --flash_freq 80m 0x1000 build/bootloader/bootloader.bin 0x8000 build/partition_table/partition-table.bin 0x10000 build/camwebsrv.bin 0x110000 build/storage.bin or run 'idf.py -p (PORT) flash'

    ```

5. Upload to the ESP32-CAM board, assuming that it is connected to `/dev/ttyUSB0`
    ```
    $ idf.py -p /dev/ttyUSB0 flash
    Executing action: flash
    ...
    Leaving...
    Hard resetting via RTS pin...
    Done
    ```

## Author

[Vino Fernando Crescini](mailto:vfcrescini@gmail.com)

## License

This software is released under the GPLv3 license. For more information, see [COPYING](https://github.com/vfcrescini/esp32-cam-websrv/blob/main/COPYING).


ChangeLog
2023-07-27  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* VERSION:

	  - Updated to 0.2.0.

	* Merged changes from esp32-cam-websrv-0.1.0.br-fix_wifi-1.


2023-07-27  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* VERSION:

	  - Updated to 0.1.0-br-fix_wifi-1.

	* main/config.h:

	  - CAMWEBSRV_PING_TIMEOUT_INIT: renamed to CAMWEBSRV_PING_TIMEOUT_SEND

	  - CAMWEBSRV_PING_TIMEOUT_SENT: renamed to CAMWEBSRV_PING_TIMEOUT_RECV

	  - CAMWEBSRV_PING_CYCLE_INTERVAL: ping interval is now 30 secs.

	* main/ping.c:

	  - Consolidated states INIT and BLCK.

	  - Renamed states INIT and SENT to SEND and RECV, respectively.

	  - Fixed off-by-one timeout counter bug.

	  - Ping success log events are now logged only at DEBUG and VERBOSE
	    log levels.

	  - If ping_host is unset, or set to an empty string, the ping module
	    will be disabled.


2023-07-26  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* main/main.c:

	  - Call camwebsrv_ping_process() from main event loop. Trigger
	    delayed restart if ping fails.

	* main/ping.c:

	  - camwebsrv_ping_process(): Fixed nextevent.

	  - camwebsrv_ping_process(): Returns without doing anything if called
	    before next event is due.


2023-07-22  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* main/config.h:

	  - Added new CAMWEBSRV_MAIN_REBOOT_DELAY_MSEC config parameter.

	* main/main.c:

	  - Will now automatically reboot after a preconfigured delay when an
	    error occurrs.

	* main/wifi.c:

	  - camwebsrv_wifi_init() now returns ESP_ERR_TIMEOUT when it times
	    out.


2023-07-20  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* main/main.c:
	* main/wifi.c:
	* main/wifi.h:

	  - Tidy up: camwebsrv_wifi_init() now gives back a handle to a wifi
	    structure (object).


2023-07-10  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* main/main.c:
	* main/wifi.c:

	  - Tidy up: moved creation of default event loop to main module.


2023-07-07  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* main/ping.c:
	* main/ping.h:

	  - Added pingger module.

	* main/CMakeLists.txt:

	  - Added lwip dependency.


2023-07-04  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* main/storage.c:

	  - Use local copy of errno.


2023-06-25  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* main/wifi.c:
	* main/main.c:

	  - Moved NVS initialisation from wifi module to main.


2023-03-06  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* main/httpd.c:
	* main/httpd.h:
	* main/sclients.c:
	* main/sclients.h:
	* main/main.c:

	  - Moved sclients event mutex to main event loop.

	  - sclients.process() now calculates how many msecs before its next
	    event is due. That is then used in the main loop block timeout.

	  - sclients.process() no longer checks against duplicate frames.
	    Each pulled frame is guaranteed unique since another check ensures
	    that the minimum time between frames has elapsed before a new one
	    is pulled.

	* main/camera.c:
	* main/camera.h:

	  - Added camwebsrv_camera_fps_get().


2023-03-01  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* VERSION:

	  - Updated to v0.1.0-br-fix_wifi.


2023-02-27  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* VERSION:

	  - Updated to v0.1.0.

	* main/httpd.c:
	* main/sclients.c:

	  - Now uses camwebsrv_vbytes_t to compose HTTP headers.


2023-02-26  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* main/config.h:

	  - Added the following definitions:

	    - CAMWEBSRV_CAMERA_FPS_MIN: 1
	    - CAMWEBSRV_CAMERA_FPS_MAX: 8
	    - CAMWEBSRV_CAMERA_DEFAULT_FPS: 4

	* main/httpd.c:
	* main/camera.c:

	  - Added new fps configuration parameter.

	* storage/ov2640.htm:
	* storage/ov3660.htm:

	  - Added new FPS slider control.


2023-02-26  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* main/config.h:

	  - Added CAMWEBSRV_CAMERA_INITIAL_FRAME_SKIP; set to 3.

	* main/camera.c:

	  - When getting a frame for the first time since init() or reset(),
	    skip the first CAMWEBSRV_CAMERA_INITIAL_FRAME_SKIP frames.

	  - When resetting, return pending frame first.

	* main/httpd.c:

	  - When resetting, disconnect all connected stream clients.

	* main/sclients.h:
	* main/sclients.c:

	  - Added camwebsrv_sclients_purge() to disconnect all connected
	    clients.


2023-02-26  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* main/camera.c:

	  - Flash state is now set back to default in
	    camwebsrv_camera_reset().


2023-02-25  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* storage/script.js:

	  - Tidied up!

	  - Reset button now stops streaming and does a status update.


2023-02-25  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* main/camera.c:
	* main/camera.h:

	  - Added new camwebsrv_camera_reset() function to deinitialise and
	    then reinitialise the camera.

	* main/httpd.c:

	  - Added http://<host>/reset handler to reset camera.

	* storage/ov2640.htm:
	* storage/ov3660.htm:
	* storage/script.js:

	  - Added new reset button to reset camera.

	* main/camera.c:
	* main/camera.h:
	* main/httpd.c:

	  - Renamed camwebsrv_camera_set() and camwebsrv_camera_get() to
	    camwebsrv_camera_ctrl_set() and camwebsrv_camera_ctrl_get(),
	    respectively.


2023-02-21  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* main/camera.c:

	  - Camera module now uses a separate mutex for frame grabbing and
	    another one for all other functions.


2023-02-19  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* storage/script.js:

	  - No need to load a new still if we are already streaming.

	* main/sclients.c:

	  - camwebsrv_sclients_add() now logs client IP and port.

	* sdkconfig.defaults:

	  - Added HTTPD_QUEUE_WORK_BLOCKING=y


2023-02-15  Vino Fernando Crescini  <vfcrescini@gmail.com>

	* storage/script.js:

	  - Requests to http://<server>/stream and http://<server>/capture are
	    now made with an extra unused random number parameter. This is
	    workaround for a problem in some browsers that prevents a second

main/config.h
// 2023-01-03 config.h
// Copyright (C) 2023 Vino Fernando Crescini  <vfcrescini@gmail.com>
// SPDX-License-Identifier: GPL-3.0-or-later

#ifndef _CAMWEBSRV_CONFIG_H
#define _CAMWEBSRV_CONFIG_H

#define CAMWEBSRV_TAG "camwebsrv"

#define CAMWEBSRV_MAIN_REBOOT_DELAY_MSEC 10000
#define CAMWEBSRV_MAIN_MIN_CYCLE_MSEC 10

#define CAMWEBSRV_CFGMAN_FILENAME "config.cfg"
#define CAMWEBSRV_CFGMAN_KEY_WIFI_SSID "wifi_ssid"
#define CAMWEBSRV_CFGMAN_KEY_WIFI_PASS "wifi_pass"
#define CAMWEBSRV_CFGMAN_KEY_PING_HOST "ping_host"

#define CAMWEBSRV_CAMERA_INITIAL_FRAME_SKIP 3
#define CAMWEBSRV_CAMERA_FPS_MIN 1
#define CAMWEBSRV_CAMERA_FPS_MAX 8
#define CAMWEBSRV_CAMERA_DEFAULT_FS 10
#define CAMWEBSRV_CAMERA_DEFAULT_FPS 4
#define CAMWEBSRV_CAMERA_DEFAULT_FLASH false

#define CAMWEBSRV_VBYTES_BSIZE 16

#define CAMWEBSRV_SCLIENTS_BSIZE 8192
#define CAMWEBSRV_SCLIENTS_SEND_TMOUT 1000
#define CAMWEBSRV_SCLIENTS_IDLE_TMOUT 3000

#define CAMWEBSRV_PING_TIMEOUT_MAX 3
#define CAMWEBSRV_PING_TIMEOUT_SEND 5000
#define CAMWEBSRV_PING_TIMEOUT_RECV 5000
#define CAMWEBSRV_PING_WAIT_INTERVAL 1000
#define CAMWEBSRV_PING_CYCLE_INTERVAL 30000

// ESP32Cam (AiThinker) PIN Map

#define CAMWEBSRV_PIN_PWDN 32
#define CAMWEBSRV_PIN_RESET -1
#define CAMWEBSRV_PIN_XCLK 0
#define CAMWEBSRV_PIN_SIOD 26
#define CAMWEBSRV_PIN_SIOC 27

#define CAMWEBSRV_PIN_D7 35
#define CAMWEBSRV_PIN_D6 34
#define CAMWEBSRV_PIN_D5 39
#define CAMWEBSRV_PIN_D4 36
#define CAMWEBSRV_PIN_D3 21
#define CAMWEBSRV_PIN_D2 19
#define CAMWEBSRV_PIN_D1 18
#define CAMWEBSRV_PIN_D0 5
#define CAMWEBSRV_PIN_VSYNC 25
#define CAMWEBSRV_PIN_HREF 23
#define CAMWEBSRV_PIN_PCLK 22

#define CAMWEBSRV_PIN_FLASH 4

#endif
